feat(compositions): Support patching from Environment

MisterMX opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Design: https://github.com/crossplane/crossplane/pull/3008

Fixes: #2099 

This adds a new cluster scope CRD `EnvironmentConfig` that is similar to a standard `ConfigMap` but allows storing arbitrary values instead of just strings.

`EnvironmentConfig`s can be used in compositions using the new `FromEnvironmentFieldPath` patch type.

The main use case of this PR is to use to build a generic Crossplane package that can run in multiple environment without having to hardcode environment specific values in the compositions.

This is a replacement for #2938. Based on an idea briefly mentioned by @negz in the community meeting.

It avoids the discussed security and RBAC issues by not requiring access to secrets and config maps and relying on Crossplane native resources instead.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Manually and unit tests

[contribution process]: https://git.io/fj2m9

negz commented on apis/apiextensions/v1alpha1/config_set_types.go:
```diff
@@ -0,0 +1,33 @@
+// A ConfigSet contains a set of arbitrary, unstructured values.
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Cluster,categories=crossplane
+type ConfigSet struct {
```
I'd like to call this an `EnvironmentConfig` rather than a `ConfigSet`, for symmetry with our existing `ProviderConfig` and `StoreConfig` types. Or more generally I'd like to call it something that ends in `Config` - I'm happy if there are better suggestions than `Environment`.

MisterMX replied:
What about `CompositionEnvironmentConfig`?

MisterMX replied:
Solved in https://github.com/crossplane/crossplane/pull/3008#discussion_r902352449

MisterMX commented:
I added a new type `EnvironmentPatch` that fits to the requirements of `spec.environment.patches` to keep the API design clean.

To ensure consistency between `Composition` and `CompositionRevision` (which was not given before) I moved all shared types like `Patch` to from `apextensions/v1` to `apiextensions/common` and removed redundant code from revisions.

sboschman commented:
I realise I am way late to the party, and a lot of work has already gone into this. Just recently started experimenting with Crossplane in the hope to come up with a solution for some of the issues we are trying to tackle in this whole provisioning dance.

At the moment I am stuck with two Crossplane use-cases that would benefit from a more generic implementation in the spirit of #1770 , than from imo a very use-case specific implementation for environment overrides with a ConfigMap clone.

**Use case 1:**
Patch a composed resource with a value from a composed resource belonging to a different XR. I have no problem with patching XR 1 with the required field from one of its composed resources, and then referencing this field from XR 1 in XR 2.

XR 1 - MyClusterXR -> composition Cluster.rancher2.rancher.jet... + ...
XR 2 - MyProjectXR -> composition Project.rancher2.rancher.jet.... + ....

Cluster.rancher2... composed resource has a field status.atProvider.id (this is the Rancher cluster Id, but there are more fields like systemProjectId and defaultProjectId that you might need to provide to other resources from the rancher provider)

Composition for MyProjectXR (snippet):
```
  resources:
    - base:
        apiVersion: rancher2.rancher.jet.crossplane.io/v1alpha1
        kind: Project
        spec:
          forProvider:
            clusterId: ?
      patches:
        - toFieldPath: spec.forProvider.clusterId
           from: <reference to the cluster id exposed by MyClusterXR - MyClusterXR-123.status.atProvider.id>
```
 
**Use case 2:**
Patch with a value from any k8s resource. We are using both Rancher and ArgoCD. To link a k8s namespace to a Rancher project the namespace requires Rancher specific labels and annotations. Generating the namespace with the Rancher API (using Crossplane) automatically sets the correct labels and annotations. But... the namespaces are only known after the ArgoCD ApplicationSet controller has generated the ArgoCD Application resources. So, issue one is to have some sort of Crossplane composed resource generator. This 'generator' has to create a composed Namespace.rancher2.rancher.jet... resource for every ArgoCD Application resource. Issue two is the ability to reference the ArgoCD Application resource and patch the composed resource with a value from it.

ArgoCD ApplicationSet (composed resource as part of an XD, as dev teams have to provide some input for it and it has to live in the argocd namespace, to which teams should not have direct access):
```
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
spec:
  generators:
  - git:
      repoURL: https://github.com/argoproj/applicationset.git
      revision: HEAD
      directories:
      - path: examples/git-generator-directory/cluster-addons/*
  template:
    spec:
      destination:
        namespace: '{{path.basename}}'
```

Short explanation, ArgoCD will watch the git repo and generate an ArgoCD Application resource for every directory it finds. Add/remove a directory and the ArgoCD Application resources are auto generated.

So, for example, if it contains the directory `argo-workflows` it generates the following ArgoCD Application resource:
```
apiVersion: argoproj.io/v1alpha1
kind: Application
spec:
  destination:
    namespace: 'argo-workflows'
```

Now ArgoCD can generate the namespace, but without the required Rancher labels and annotations. Also, ArgoCD will not delete the namespace, as it follows the Helm way of '--create-namespace' flag. Especially annoying in a cluster serving pull requests, as you end up with quite a few 'empty' namespaces for obsolete pull requests. So, if you have other operators running that do their thing for every namespace... it starts to hurt.

So, I was looking into Crossplane to take over the namespace creation (and deletion) using the Rancher API. Creating some sort of composition like:
```
  resources:
    - base:
        apiVersion: rancher2.rancher.jet.crossplane.io/v1alpha1
        kind: Namespace
        metadata:
          name: <value somehow provided by Application.argoproj.io/v1alpha1 spec.destination.namespace>
        spec:
          forProvider:
            projectId: <value somehow referenced from the Project XR, use case 1>
```

I realise, for this to work, it requires another Crossplane feat. As I don't think it is currently possible to generate multiple composed resources from a single composed resource template. But that part is for another issue/feat request.

nm-morais commented:
Hey @MisterMX! Great work putting this together 💪! I'm wondering. is this PR at a stage where we could checkout and test the feature?

MisterMX commented:
@nm-morais this PR is just a draft to serve as a demo for the design discussion in #3008. Although it has been updated for some time you should be able to build and run it.

I will come back to this PR once the design discussion is finished.

MisterMX commented:
I refactored the branch, rebased on the latest master and changed the API to match the design document from #3008. I also added more unit tests for selecting and fetching EnvironmentConfigs.

These changes require https://github.com/crossplane/crossplane-runtime/pull/353 to be merged before we can continue here. The code iteself, however, is ready for review.

negz commented:
@MisterMX I'd like to get someone else to own at least the first pass review of this implementation (it will probably go faster that way). GitHub should assign reviewers once you mark it ready for review. I've just approved the runtime PR that was blocking it.

ytsarev commented on apis/apiextensions/v1/composition_patches.go:
```diff
@@ -46,11 +46,13 @@ type PatchType string
+	PatchTypeFromEnvironmentFieldPath PatchType = "FromEnvironmentFieldPath"
+	PatchTypePatchSet                 PatchType = "PatchSet"
+	PatchTypeToCompositeFieldPath     PatchType = "ToCompositeFieldPath"
+	PatchTypeCombineFromEnvironment   PatchType = "CombineFromEnvironment"
```
Similar to the above, we need to extend associated API field documentation at https://github.com/crossplane/crossplane/pull/3007/files#diff-fdd289c27b8b59d09910862c76402dd0be37b082140e139f9b36afeb6e987a21R97-R100 with something like:

```
    // Combine is the patch configuration for a CombineFromComposite, CombineFromEnvironment or
    // CombineToComposite patch.
    // +optional
    Combine *Combine `json:"combine,omitempty"`
```

MisterMX replied:
Added tests to `apis/apiextensions/v1/composition_patches_test.go` and updated documentation.

ytsarev reviewed (changes requested):
@MisterMX Thanks a lot for this impressive contribution! It is a highly-anticipated feature.

I've left some inline questions and suggestions please check them out.

I've also tried to test the feature end-to-end.

I am running the Crossplane controller locally with this feature enabled

```
$(GO_OUT_DIR)/$(PROJECT_NAME) core start --debug --enable-composition-environments
```

New and updated CRDs are also applied

```
285     0       cluster/crds/apiextensions.crossplane.io_compositions.yaml
56      0       cluster/crds/apiextensions.crossplane.io_environmentconfigs.yaml
```

I created the following set of XRD+Composition+EnvironmentConfig+XR for the test https://github.com/ytsarev/composition-environment-example

I believe I have machine EnvironmentConfig in the Composition:

```
    environment:
      environmentConfigs:
        - ref:
            name: example-environment
```

And associated EnvironmentConfig contains the `region: eu-central-1` field 

Still, I am getting `no such field` when creating XR( I've also tried to reference the field as `data.region` in https://github.com/ytsarev/composition-environment-example/blob/main/composition.yaml#L27 , same result) 

```
Spec:
  Composition Ref:
    Name:                     xvpcs.aws.platformref.upbound.io
  Composition Update Policy:  Automatic
  Resource Refs:
    API Version:  ec2.aws.upbound.io/v1beta1
    Kind:         VPC
  Vpc Name:       environment-config-test
Status:
  Conditions:
    Last Transition Time:  2022-11-09T02:27:29Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced
    Last Transition Time:  2022-11-09T02:27:29Z
    Reason:                Creating
    Status:                False
    Type:                  Ready
Events:
  Type     Reason             Age              From                                                             Message
  ----     ------             ----             ----                                                             -------
  Normal   SelectComposition  2s (x4 over 3s)  defined/compositeresourcedefinition.apiextensions.crossplane.io  Successfully selected composition
  Warning  ComposeResources   2s (x4 over 3s)  defined/compositeresourcedefinition.apiextensions.crossplane.io  cannot render composed resource from resource template at index 0: cannot apply the patch at index 1: region: no such field
  Normal   ComposeResources   2s (x4 over 3s)  defined/compositeresourcedefinition.apiextensions.crossplane.io  Successfully composed resources
```

I am generally missing `environmentConfigRefs` on XR level that are mentioned in the design doc https://github.com/crossplane/crossplane/blob/master/design/one-pager-composition-environment.md?plain=1#L147-L151, so not sure if the environment is getting compiled.

Unfortunately, PR does not contain any examples, so I might be missing something by compiling the specs with the help of a design doc and `kubectl explain`.

Is there anything missing in my test configuration?

MisterMX commented:
@ytsarev thanks for the review.

> Still, I am getting no such field when creating XR( I've also tried to reference the field as data.region in [ytsarev/composition-environment-example@main/composition.yaml#L27](https://github.com/ytsarev/composition-environment-example/blob/main/composition.yaml?rgh-link-date=2022-11-09T02%3A37%3A04Z#L27) , same result)

The `data` path is prepended automatically. So in your case the path should be just `region`.

ytsarev commented:
> The data path is prepended automatically. So in your case the path should be just region.

Cool so it is matching my original attempt https://github.com/ytsarev/composition-environment-example/blob/main/composition.yaml#L27 , still it reports missing the field.

MisterMX commented on apis/apiextensions/v1/composition_environment.go:
```diff
@@ -0,0 +1,217 @@
+	// Ref is a named reference to a single EnvironmentConfig.
+	// Either Ref or Selector is required.
+	// +optional
+	Ref *EnvironmentSourceReference `json:"ref,omitempty"`
```
I left it out, because it is implied implicitly by lying inside `EnvironmentSource`:

```yaml
spec:
  environment:
    environmentConfigs:
    - type: Reference
      reference:
        name: example-environment
```

ytsarev commented on apis/apiextensions/v1/composition_environment.go:
```diff
@@ -0,0 +1,217 @@
+type EnvironmentConfiguration struct {
+	// EnvironmentConfigs selects a list of `EnvironmentConfig`s. The resolved
+	// resources are stored in the composite resource at
+	// `spec.environmentConfigRefs` and is only updated if it is null.
```
Thanks, I've double-checked the associated crossplane-runtime code at https://github.com/crossplane/crossplane-runtime/pull/353/files#diff-abbe009bdd585ac4b3617965597880a55ac19e0c3f5b4eb65179cb9bd4edbb74R241

It is confusing, why does it set `spec.resourceRefs` and not `spec.environmentConfigRefs` ?

MisterMX replied:
Yes, that was a mistake. It was in https://github.com/crossplane/crossplane-runtime/pull/365.

MisterMX commented on internal/controller/apiextensions/composite/composed.go:
```diff
@@ -348,6 +349,11 @@ func (r *APIDryRunRenderer) Render(ctx context.Context, cp resource.Composite, c
 		if err := t.Patches[i].Apply(cp, cd, patchTypesFromXR()...); err != nil {
 			return errors.Wrapf(err, errFmtPatch, i)
 		}
+		if env != nil {
+			if err := t.Patches[i].Apply(env, cd, patchTypesFromEnvironment()...); err != nil {
+				return errors.Wrapf(err, errFmtPatch, i)
+			}
+		}
```
There is no repetition in this part: First you are running the regular apply with composite and composed (`cp` and `cd`) and in the second step you are applying the environment patch (`env` and `cd`).

MisterMX commented:
Yes, there were some bugs with storing the EnvironmentConfig refs in the composite. It should work now.

MisterMX commented on cmd/crossplane/core/core.go:
```diff
@@ -110,6 +111,10 @@ func (c *startCommand) Run(s *runtime.Scheme, log logging.Logger) error { //noli
 	}
+	if c.EnableCompositionEnvironments {
+		feats.Enable(features.EnableAlphaCompositionEnvironment)
+		log.Info("Alpha feature enabled", "flag", features.EnableAlphaCompositionEnvironment)
```
I added a hint to the design doc. Though I think it should be noted in the release notes as well.

MisterMX commented on cmd/crossplane/core/core.go:
```diff
@@ -73,8 +73,9 @@ type startCommand struct {
-	EnableCompositionRevisions bool `group:"Alpha Features:" help:"Enable support for CompositionRevisions."`
-	EnableExternalSecretStores bool `group:"Alpha Features:" help:"Enable support for ExternalSecretStores."`
+	EnableCompositionRevisions    bool `group:"Alpha Features:" help:"Enable support for CompositionRevisions."`
+	EnableCompositionEnvironments bool `group:"Alpha Features:" help:"Enable support for Composition environments."`
```
Renamed the flag to `--enable-environment-configs`

MisterMX commented on apis/apiextensions/v1/composition_patches.go:
```diff
@@ -46,11 +46,13 @@ type PatchType string
+	PatchTypeToCompositeFieldPath     PatchType = "ToCompositeFieldPath"
+	PatchTypeCombineFromEnvironment   PatchType = "CombineFromEnvironment"
+	PatchTypeCombineFromComposite     PatchType = "CombineFromComposite"
+	PatchTypeCombineToComposite       PatchType = "CombineToComposite"
```
Whoops. Seems like I missed them. I added the implementation for them as well.

ytsarev reviewed (commented):
@MisterMX could you please check the pipes before I jump into the 2nd round of review? :)

ytsarev reviewed (changes requested):
@MisterMX thanks a ton for the fixes.

Please check the inline comments especially the one around `panic` 

Also I've e2e tested `FromEnvironmentFieldPath` with the example from https://github.com/ytsarev/composition-environment-example/blob/main/composition.yaml#L26-L30 , it works properly now 👍 

I've also tried to test `ToEnvironmentFieldPath` , here is the attempt
https://github.com/ytsarev/composition-environment-example/commit/fdc27a24355d91e9da5c6dfeb319ebe550dce868
It does not seem to work - I can't get vpcId value from in-memory Environment. Am I using it as expected? 

In this context, some documentation with example usage of this functionality is visibly required. Do you mind extending the reference documentation at `docs/reference/composition.md`?

MisterMX commented on internal/controller/apiextensions/composite/reconciler.go:
```diff
@@ -505,6 +592,16 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
+	// If we have an environment, run all environment patches before composing
+	// resources.
+	if env != nil {
+		for i, p := range comp.Spec.Environment.Patches {
```
Fixed

MisterMX commented on apis/apiextensions/v1/composition_patches.go:
```diff
@@ -47,11 +47,15 @@ type PatchType string
+	PatchTypeFromEnvironmentFieldPath PatchType = "FromEnvironmentFieldPath"
+	PatchTypePatchSet                 PatchType = "PatchSet"
+	PatchTypeToCompositeFieldPath     PatchType = "ToCompositeFieldPath"
+	PatchTypeToEnvironmentFieldPath   PatchType = "ToEnvironmentFieldPath"
```
Added tests

MisterMX commented on apis/apiextensions/v1/composition_patches.go:
```diff
@@ -47,11 +47,15 @@ type PatchType string
+	PatchTypeCombineFromEnvironment   PatchType = "CombineFromEnvironment"
+	PatchTypeCombineFromComposite     PatchType = "CombineFromComposite"
+	PatchTypeCombineToComposite       PatchType = "CombineToComposite"
+	PatchTypeCombineToEnvironment     PatchType = "CombineToEnvironment"
```
Added tests

MisterMX commented:
@ytsarev thanks for looking into it.

> I've also tried to test ToEnvironmentFieldPath , here is the attempt
https://github.com/ytsarev/composition-environment-example/commit/fdc27a24355d91e9da5c6dfeb319ebe550dce868
It does not seem to work - I can't get vpcId value from in-memory Environment. Am I using it as expected?

The reason patching from `status` does not work is because the patches are applied *before* the resource is created / updated on the cluster. And until then the `status` field remains empty.

`ToComposite*` patches are applied out-of-order after the resource is created on the cluster and the status information is received. That's why they work for this. (This is an undocumented behaviour btw, see #3454).

ytsarev reviewed (changes requested):
Can't build this branch anymore, apparently the same happens in the pipelines. 

```
make run
22:12:29 [ .. ] go build darwin_arm64
go: -i flag is deprecated
github.com/crossplane/crossplane/apis/apiextensions/v1alpha1
# github.com/crossplane/crossplane/apis/apiextensions/v1alpha1
apis/apiextensions/v1alpha1/zz_generated.deepcopy.go:287:29: undefined: v1.JSON
```
@MisterMX Please check

MisterMX commented:
@ytsarev it should work now. I was missing a `make generate`. Can you check again?

ytsarev commented:
@MisterMX looks like pipes are failing after recent force-push update.

In addition to the Composition Reference doc extension could you please drop some e2e examples? 

Similar to https://github.com/ytsarev/composition-environment-example but derived from your working samples so the use case and the style of consumption of this feature will be totally clear.

MisterMX commented:
@ytsarev I added a sample to `docs/snippets/composition-environments/`.

ytsarev commented:
@MisterMX amazing, thank you! going to retest it soon!

ytsarev commented on docs/snippets/composition-environments/composition.yaml:
```diff
@@ -0,0 +1,52 @@
+            cidrBlock: 172.16.0.0/16
+      patches:
+        - fromFieldPath: spec.vpcName
+          toFieldPath: spec.forProvider.tags.Name
```
If we are using comminity provider we need to change patch style https://marketplace.upbound.io/providers/crossplane-contrib/provider-aws/v0.33.0/resources/ec2.aws.crossplane.io/VPC/v1beta1#doc:spec-forProvider-tags

Otherwise it fails with
```
  Warning  ComposeResources   4m3s                    defined/compositeresourcedefinition.apiextensions.crossplane.io  cannot render composed resource from resource template at index 0: cannot use dry-run create to name composed resource: VPC.ec2.aws.crossplane.io "vpc-using-environment-config-wzxtw" is invalid: spec.forProvider.tags: Invalid value: "object": spec.forProvider.tags in body must be of type array: "object"
```

The current form will work with the official provider though as we have `tags` as `object` there https://marketplace.upbound.io/providers/upbound/provider-aws/v0.24.0/resources/ec2.aws.upbound.io/VPC/v1beta1#doc:spec-forProvider-tags

MisterMX replied:
Thanks for the hint. I switched to the community provider and changed `tag` to an array.

ytsarev commented on docs/snippets/composition-environments/composition.yaml:
```diff
@@ -0,0 +1,52 @@
+          toFieldPath: vpcId
+    - name: subnet
+      base:
+        apiVersion: ec2.aws.upbound.io/v1beta1
```
it is the official provider in this case.

I do not have a strong preference which provider-aws to use in the snippet - official or community one - but let's not mix two of them in a single Composition

MisterMX replied:
Using only community provider now.

ytsarev commented on docs/snippets/composition-environments/xrd.yaml:
```diff
@@ -0,0 +1,25 @@
+              vpcName:
+                type: string
+            required:
+            - vpcName
```
Please also add file with XR https://raw.githubusercontent.com/ytsarev/composition-environment-example/main/xvpc.yaml for easy end-to-end testing

MisterMX replied:
Added a `claim.yaml`

ytsarev commented on internal/controller/apiextensions/composite/composite.go:
```diff
@@ -27,3 +27,13 @@ func patchTypesToXR() []v1.PatchType {
 	return []v1.PatchType{v1.PatchTypeFromCompositeFieldPath, v1.PatchTypeCombineFromComposite}
 }
+
+// Returns types of patches that are _from_ the environment to a composed resource.
```
Should it contain `ToEnvironment` patch types? If yes, please amend the comment

MisterMX replied:
Changed comment.

ytsarev reviewed (changes requested):
@MisterMX example snippets require some fixes. It would be also great to have a good example of one of `ToEnvironment` patch types.

MisterMX commented on docs/snippets/composition-environments/composition.yaml:
```diff
@@ -0,0 +1,52 @@
+          toFieldPath: spec.forProvider.region
+          policy:
+            fromFieldPath: Required
+        - type: ToEnvironmentFieldPath
```
Now we do :)

ytsarev commented on docs/snippets/composition-environments/composition.yaml:
```diff
@@ -0,0 +1,60 @@
+            fromFieldPath: Required
+        - type: FromEnvironmentFieldPath
+          fromFieldPath: vpcId
+          toFieldPath: spec.forProvider.vpcId
```
we have an issue here

```
  Warning  CannotCreateExternalResource     86s (x24 over 104s)  managed/subnet.ec2.aws.crossplane.io  failed to create the Subnet resource: api error InvalidVpcID.NotFound: The vpc ID 'us-west-1' does not exist
```

MisterMX replied:
Fixed

ytsarev commented on docs/snippets/composition-environments/composition.yaml:
```diff
@@ -0,0 +1,60 @@
+          policy:
+            fromFieldPath: Required
+        - type: ToEnvironmentFieldPath
+          fromFieldPath: spec.forProvider.region
```
this usage will break below in Subnet

MisterMX replied:
Fixed

ytsarev commented on docs/snippets/composition-environments/composition.yaml:
```diff
@@ -0,0 +1,60 @@
+          policy:
+            fromFieldPath: Required
+        - type: ToEnvironmentFieldPath
+          fromFieldPath: spec.forProvider.region
+          toFieldPath: vpcId
```
```suggestion
          fromFieldPath: spec.forProvider.cidrBlock
          toFieldPath: cidrBlock
```

MisterMX replied:
Fixed

ytsarev commented on docs/snippets/composition-environments/composition.yaml:
```diff
@@ -0,0 +1,60 @@
+        kind: Subnet
+        spec:
+          forProvider:
+            cidrBlock: 172.16.20.0/24
```
```suggestion
            vpcIdSelector:
               matchControllerRef: true
```

MisterMX replied:
Fixed

ytsarev commented on docs/snippets/composition-environments/composition.yaml:
```diff
@@ -0,0 +1,60 @@
+          policy:
+            fromFieldPath: Required
+        - type: FromEnvironmentFieldPath
+          fromFieldPath: vpcId
+          toFieldPath: spec.forProvider.vpcId
```
```suggestion
          fromFieldPath: cidrBlock
          toFieldPath: spec.forProvider.cidrBlock
```

MisterMX replied:
Fixed

ytsarev reviewed (changes requested):
@MisterMX Thanks a lot for the fixes, hopefully, we are really close to the last iteration! :) 

The test composition has issues, I've left inline comments and suggestions there.

Overall looks great after fixing the example Composition 👍 
```
k get -f claim.yaml
NAME               SYNCED   READY   CONNECTION-SECRET   AGE
environment-test   True     True                        8m28s

k get managed
NAME                                                     READY   SYNCED   ID                      CIDR            IPV6CIDR   AGE
vpc.ec2.aws.crossplane.io/environment-test-dvq5k-lvxwd   True    True     vpc-0ba754df1c49197ff   172.16.0.0/16              8m52s

NAME                                                        READY   SYNCED   ID                         VPC                     CIDR            AGE
subnet.ec2.aws.crossplane.io/environment-test-dvq5k-fvwfx   True    True     subnet-01a494e14bf5458b5   vpc-0ba754df1c49197ff   172.16.0.0/16   8m52s
```

ytsarev reviewed (approved):
Everything worked and looks great to me! Super happy to approve this amazing change, thanks a lot @MisterMX!

Just a quick reference for the future on the local test steps I performed before the final approval.

1. Modify the `Makefile` `run:` to enable this alpha feature
```
-       $(GO_OUT_DIR)/$(PROJECT_NAME) core start --debug
+       $(GO_OUT_DIR)/$(PROJECT_NAME) core start --debug --enable-environment-configs
```

2. `k apply -f ./cluster/crds/` and `make run` to execute crossplane locally. Don't forget to pause Crossplane/UXP with `k -n upbound-system scale deploy crossplane --replicas=0` to avoid parallel controller running.
3. Apply test examples
``` 
cd docs/snippets/composition-environments
k apply -f environment.yaml,composition.yaml,xrd.yaml
k apply -f claim.yaml
```
4. Validate the results:
```
k get -f claim.yaml
NAME               SYNCED   READY   CONNECTION-SECRET   AGE
environment-test   True     True                        48s
k get managed
NAME                                                        READY   SYNCED   ID                         VPC                     CIDR            AGE
subnet.ec2.aws.crossplane.io/environment-test-mjbx7-h6mpx   True    True     subnet-0ea69d13d8841ef20   vpc-0e945063573920a1a   172.16.0.0/16   76s

NAME                                                     READY   SYNCED   ID                      CIDR            IPV6CIDR   AGE
vpc.ec2.aws.crossplane.io/environment-test-mjbx7-md4wm   True    True     vpc-0e945063573920a1a   172.16.0.0/16              76s
```

turkenh reviewed (approved):
This is a great feature @MisterMX, kudos for making it to the finish line 🙌 
I haven't done a thorough review, but I rely on @ytsarev for that.
